Фамильное дерево. Начало

FominaCathy opened the pull request.

FominaCathy committed:
фамильное дерево. начало.

FominaCathy committed:
фамильное дерево. начало.

FominaCathy committed:
фамильное дерево. начало.

FominaCathy committed:
фамильное дерево. начало.

Liberate520 commented on src/main/FamilyTree/FamilyTree.java:
```diff
@@ -0,0 +1,234 @@
+    }
+
+    //TODO нужно подумать
+    public void addChildren(int innParent, Human child) {
+        Human parent = getHumanForINN(innParent);
+        parent.addChildren(child);
+        addHuman(child);
+    }
```
Лучше создавать обоюдные связи, чтобы не было путаницы и было подобие автоматизации

FominaCathy replied:
переделала структуру дерева

Liberate520 commented on src/main/FamilyTree/FamilyTree.java:
```diff
@@ -0,0 +1,234 @@
+     *
+     * @return
+     */
+    public List<String> getAllFamilyTree() {
```
Интересный конечно метод, но можно было просто вызвать toString у листа

FominaCathy replied:
не хочу возвращать список ссылок на элементы дерева. чтобы не было возможности изменить элементы не через работу с деревом

Liberate520 commented on src/main/FamilyTree/FamilyTree.java:
```diff
@@ -0,0 +1,234 @@
+    public List<String> getChildren(int innHuman) {
+        return familyTree.stream()
+                .filter(e -> e.getInn() == innHuman)
+                .toList().get(0)
```
есть findFirst

FominaCathy committed:
добавлены: методы записи в файл и чтения из файла, реализован  интерфейс Iterable для дерева, методы сортировки: по имени и по возрасту

FominaCathy committed:
добавлена пакетная структура

FominaCathy commented:
- добавила методы записи в файл и чтения из файла, 
- реализован интерфейс Iterable для дерева, 
- добавила методы сортировки: по имени и по возрасту
- добавила пакетную структуру

Liberate520 commented on src/main/FamilyTree/Tree/FileHandler.java:
```diff
@@ -0,0 +1,33 @@
+package FamilyTree.Tree;
```
Класс не должен быть частью пакета описывающего дерево

FominaCathy replied:
Убрала из пакета дерева

FominaCathy replied:
поправила

Liberate520 commented on src/main/FamilyTree/Main.java:
```diff
@@ -0,0 +1,92 @@
+        print(familyTree, "\n----- первоначальный список");
+
+        //сортировки
+        sortedPrint(familyTree);
```
Методы сортировки, по заданию, должны были быть в классе дерева

FominaCathy replied:
поправила

FominaCathy committed:
1. Добавила интерфейс Person. Класс Human имплеметирует интерфейс Person

2. изменила классы компараторы.
теперь они для Интерфейса Person (не знаю можно ли так делать?)

3. Класс FamilyTree теперь параметризированный тип
и принимает классы реализующие интерфейс Person

4. изменила структуру пакетов: у меня есть пакет для дерева, и пакет для элемента дерева

FominaCathy committed:
изменила структуру фамильного дерева.

FominaCathy commented:
1. Добавила интерфейс Person. Класс Human реализует интерфейс Person

2. изменила классы компараторы.
теперь они для Интерфейса Person (не знаю можно ли так делать?)

3. Класс FamilyTree теперь параметризированный тип
и принимает классы реализующие интерфейс Person

4. изменила структуру пакетов: у меня есть пакет для дерева, и пакет для элемента дерева

FominaCathy committed:
Реализовала паттерн MVP

FominaCathy commented:
добавила паттерн MVP

Liberate520 commented on src/main/FamilyTree/Model/Service/Service.java:
```diff
@@ -0,0 +1,73 @@
+
+public class Service {
+    private FamilyTree<Human> tree;
+    private FileHandler fileHandler;
```
нарушает пятый принцип солид. Зависимость должна быть через интерфейс

FominaCathy replied:
исправила

Liberate520 commented on src/main/FamilyTree/View/ConsoleUI.java:
```diff
@@ -0,0 +1,252 @@
+            int innParent = -1;
+            int innChildren = -1;
+
+            System.out.println("Выберете вариант добавления:\n" +
+                    "1. Оба есть в дереве -> добавление по ИНН\n" +
+                    "2. В дереве есть только родитель -> \n" +
+                    "3. В дереве есть только ребенок ->\n" +
+                    "0. Выход в главное меню \n");
```
Возможно нарушение второго принципа солид. Проект в этой части не может быть расширен, но возможно вы предполагаете, что это никогда не потребуется

FominaCathy replied:
Исправила: добавила 2 подменю.

FominaCathy committed:
Добавива 2 подменю.
исправила класс Service (теперь там завииость через интерфейс с Writable)

FominaCathy commented:
добавила 2 подменю и исправила зависимость в сервисе

Liberate520 commented on src/main/FamilyTree/Model/Human.java:
```diff
@@ -0,0 +1,192 @@
+        this.children = new ArrayList<>();
+    }
+
+    public static Human create(String firstName, String lastName, Gender gender, LocalDate birtDay) {
+        if (!birtDay.isBefore(LocalDate.now())) {
+            throw new IllegalArgumentException("некорректная дата рождения");
+        }
+        return new Human(firstName, lastName, gender, birtDay);
+    }
```
Чем это отличается от простого вызова конструктора класса? Тем, что здесь выбрасывается исключение? Но это можно было сделать и в конструкторе, если очень хочется. Зачем эта статика я не понял(

Liberate520 commented on src/main/FamilyTree/Presentor/Presenter.java:
```diff
@@ -0,0 +1,116 @@
+
+    public void beginTree(String fName, String lName, Gender gender, LocalDate dataB) {
+
+        service.beginTree(Human.create(fName, lName, gender, dataB));
```
Презентер не должен создавать людей. Человек это объект модели, вот пусть модель его и создает (в нашем случае Сервис)

Liberate520 commented on src/main/FamilyTree/View/ConsoleUI.java:
```diff
@@ -0,0 +1,256 @@
+        int innChildren = getInnPerson("ИНН ребенка");
+        System.out.println("Введите данные родителя:");
+        Map<String, Object> parent = getDataPerson();
+        presenter.establishPaternity((String) parent.get("fName"), (String) parent.get("lName"),
+                (Gender) parent.get("gender"), (LocalDate) parent.get("dataB"), innChildren);
```
Приведение типов может уронить ваше приложение. Есть более аккуратные методы, например String.valueOf или LocalDate.parse (если строка) и LocalDate.of (если числа)
